File: backend/src/main/java/com/picsou/mcp/Scopes.java
package com.picsou.mcp;

import java.util.Set;

/**
 * The complete vocabulary of access-key scopes ({@code domain:action}).
 *
 * <p>An access-key carries a subset of {@link #ALL}; each MCP {@code @Tool} method is gated by
 * exactly one of these via {@code @RequiresScope}. {@link #ALL} is the security allowlist —
 * {@code AccessKeyController} rejects any requested scope outside it, so a key can never be
 * granted a permission that no tool honours.
 *
 * <p>Read scopes end in {@code :read}; everything else mutates ({@code :write} / {@code :trigger}).
 * There is intentionally NO scope for credentials, MFA, admin settings, member management, or
 * data export — those operations have no tool, so no scope can reach them.
 */
public final class Scopes {

    private Scopes() {}

    // ── Read ────────────────────────────────────────────────────────────────
    public static final String ACCOUNTS_READ = "accounts:read";
    public static final String TRANSACTIONS_READ = "transactions:read";
    public static final String GOALS_READ = "goals:read";
    public static final String DASHBOARD_READ = "dashboard:read";
    public static final String PRICES_READ = "prices:read";
    public static final String FAMILY_READ = "family:read";

    // ── Write / trigger ───────────────────────────────────────────────────────
    public static final String ACCOUNTS_WRITE = "accounts:write";
    public static final String TRANSACTIONS_WRITE = "transactions:write";
    public static final String GOALS_WRITE = "goals:write";
    public static final String SYNC_TRIGGER = "sync:trigger";

    /** Immutable allowlist of every valid scope. Used to validate key-creation requests. */
    public static final Set<String> ALL = Set.of(
        ACCOUNTS_READ, TRANSACTIONS_READ, GOALS_READ, DASHBOARD_READ, PRICES_READ, FAMILY_READ,
        ACCOUNTS_WRITE, TRANSACTIONS_WRITE, GOALS_WRITE, SYNC_TRIGGER
    );
}

File: frontend/src/features/accessKeys/scopes.ts
/**
 * The access-key scope vocabulary, mirroring the backend allowlist
 * `com.picsou.mcp.Scopes#ALL`. The UI must offer exactly these — a scope the
 * backend doesn't recognise would make key creation fail with HTTP 400.
 *
 * Read scopes end in `:read`; everything else (`:write` / `:trigger`) mutates.
 */
export const ALL_SCOPES = [
  'accounts:read',
  'transactions:read',
  'goals:read',
  'dashboard:read',
  'prices:read',
  'family:read',
  'accounts:write',
  'transactions:write',
  'goals:write',
  'sync:trigger',
] as const

export type Scope = (typeof ALL_SCOPES)[number]

export type ScopeGroup = 'read' | 'write'

/** Classifies a scope: `:read` → read, anything else (`:write` / `:trigger`) → write. */
export function scopeGroup(scope: string): ScopeGroup {
  return scope.endsWith(':read') ? 'read' : 'write'
}

/** i18n-safe key for a scope: `accounts:read` → `accounts_read`. */
export function scopeI18nKey(scope: string): string {
  return scope.replace(':', '_')
}

export const READ_SCOPES: Scope[] = ALL_SCOPES.filter((s) => scopeGroup(s) === 'read')
export const WRITE_SCOPES: Scope[] = ALL_SCOPES.filter((s) => scopeGroup(s) === 'write')
